xds/internal/server/rds_handler_test.go

```diff
@@ -0,0 +1,423 @@
+	}
+	if gotRoute != route1 {
+		t.Fatalf("xdsClient.WatchRDS called for route: %v, want %v", gotRoute, route1)
+	}
+	rdsUpdate := xdsclient.RouteConfigUpdate{}
+	// Invoke callback with the xds client with a certain route update. Due to
+	// this route update updating every route name that rds handler handles,
+	// this should write to the update channel to send to the listener.
+	fakeClient.InvokeWatchRouteConfigCallback(route1, rdsUpdate, nil)
+	rhuWant := map[string]xdsclient.RouteConfigUpdate{route1: rdsUpdate}
+	select {
+	case rhu := <-rh.updateChannel:
+		if diff := cmp.Diff(rhu.updates, rhuWant); diff != "" {
+			t.Fatalf("got unexpected route update, diff (-got, +want): %v", diff)
+		}
+	case <-ctx.Done():
+		t.Fatal("Timed out waiting for update from update channel.")
+	}
+	// Close the rds handler. This is meant to be called when the lis wrapper is
+	// closed, and the call should cancel all the watches present (for this
+	// test, a single watch).
+	rh.close()
+	routeNameDeleted, err := fakeClient.WaitForCancelRouteConfigWatch(ctx)
+	if err != nil {
+		t.Fatalf("xdsClient.CancelRDS failed with error: %v", err)
+	}
+	if routeNameDeleted != route1 {
+		t.Fatalf("xdsClient.CancelRDS called for route %v, want %v", routeNameDeleted, route1)
+	}
+}
+
+// TestSuccessCaseTwoUpdates tests the case where the rds handler receives an
+// update with a single Route, then receives a second update with two routes.
+// The handler should start a watch for the added route, and if received a RDS
+// update for that route it should send an update with both RDS updates present.
+func (s) TestSuccessCaseTwoUpdates(t *testing.T) {
+	rh, fakeClient := setupTests(t)
+
+	routeNames := map[string]bool{route1: true}
+	rh.updateRouteNamesToWatch(routeNames)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	_, err := fakeClient.WaitForWatchRouteConfig(ctx)
+	if err != nil {
+		t.Fatalf("xdsClient.WatchRDS failed with error: %v", err)
+	}
+
+	// Update the RDSHandler with route names which adds a route name to watch.
+	// This should trigger the RDSHandler to start a watch for the added route
+	// name to watch.
+	routeNames = map[string]bool{route1: true, route2: true}
+	rh.updateRouteNamesToWatch(routeNames)
+	gotRoute, err := fakeClient.WaitForWatchRouteConfig(ctx)
+	if err != nil {
+		t.Fatalf("xdsClient.WatchRDS failed with error: %v", err)
+	}
+	if gotRoute != route2 {
+		t.Fatalf("xdsClient.WatchRDS called for route: %v, want %v", gotRoute, route2)
+	}
+
+	// Invoke the callback with an update for route 1. This shouldn't cause the
+	// handler to write an update, as it has not received RouteConfigurations
+	// for every RouteName.
+	rdsUpdate1 := xdsclient.RouteConfigUpdate{}
+	fakeClient.InvokeWatchRouteConfigCallback(route1, rdsUpdate1, nil)
+
+	// The RDS Handler should not send an update.
+	shouldNotHappenCtx, shouldNotHappenCtxCancel := context.WithTimeout(context.Background(), defaultTestShortTimeout)
```

Comment:
Really big variable names for variables with really small scope. :(

I would prefer sCtx, sCancel